File: FastOS/kernel/src/arch/x86_64/gdt.rs
// ============================================================
// FastOS — Global Descriptor Table (GDT)
// ============================================================
// Sets up GDT with null, kernel code64, kernel data64,
// user code64, user data64, and TSS segments.
// ============================================================

use super::cpu::{GdtDescriptor, lgdt};

/// GDT entry (8 bytes)
#[derive(Clone, Copy)]
#[repr(C, packed)]
pub struct GdtEntry {
    limit_low: u16,
    base_low: u16,
    base_mid: u8,
    access: u8,
    granularity: u8,
    base_high: u8,
}

impl GdtEntry {
    const fn null() -> Self {
        GdtEntry {
            limit_low: 0, base_low: 0, base_mid: 0,
            access: 0, granularity: 0, base_high: 0,
        }
    }

    const fn new(access: u8, granularity: u8) -> Self {
        GdtEntry {
            limit_low: 0xFFFF,
            base_low: 0,
            base_mid: 0,
            access,
            granularity,
            base_high: 0,
        }
    }
}

/// TSS entry (16 bytes in 64-bit mode)
#[repr(C, packed)]
pub struct TssEntry {
    length: u16,
    base_low: u16,
    base_mid: u8,
    flags1: u8,
    flags2: u8,
    base_high: u8,
    base_upper: u32,
    _reserved: u32,
}

/// Task State Segment (104 bytes minimum)
#[repr(C, packed)]
pub struct Tss {
    _reserved0: u32,
    pub rsp0: u64,       // Kernel stack for Ring 0
    pub rsp1: u64,
    pub rsp2: u64,
    _reserved1: u64,
    pub ist1: u64,       // Interrupt Stack Table entry 1 (double fault)
    pub ist2: u64,
    pub ist3: u64,
    pub ist4: u64,
    pub ist5: u64,
    pub ist6: u64,
    pub ist7: u64,
    _reserved2: u64,
    _reserved3: u16,
    pub iomap_base: u16,
}

impl Tss {
    pub const fn new() -> Self {
        Tss {
            _reserved0: 0,
            rsp0: 0, rsp1: 0, rsp2: 0,
            _reserved1: 0,
            ist1: 0, ist2: 0, ist3: 0, ist4: 0,
            ist5: 0, ist6: 0, ist7: 0,
            _reserved2: 0,
            _reserved3: 0,
            iomap_base: core::mem::size_of::<Tss>() as u16,
        }
    }
}

// GDT with 5 entries + TSS (takes 2 slots)
// Total: 7 GDT slots = 56 bytes
const GDT_ENTRIES: usize = 7;

static mut GDT: [GdtEntry; GDT_ENTRIES] = [GdtEntry::null(); GDT_ENTRIES];
static mut TSS: Tss = Tss::new();
static mut GDTR: GdtDescriptor = GdtDescriptor { limit: 0, base: 0 };

// Segment selectors (index * 8)
pub const KERNEL_CODE_SELECTOR: u16 = 1 * 8;       // 0x08
pub const KERNEL_DATA_SELECTOR: u16 = 2 * 8;       // 0x10
pub const USER_CODE_SELECTOR: u16   = 3 * 8 | 3;   // 0x1B (RPL=3)
pub const USER_DATA_SELECTOR: u16   = 4 * 8 | 3;   // 0x23 (RPL=3)
pub const TSS_SELECTOR: u16         = 5 * 8;        // 0x28

/// Initialize GDT with kernel/user segments and TSS
pub fn init() {
    unsafe {
        // Entry 0: Null
        GDT[0] = GdtEntry::null();

        // Entry 1: Kernel Code 64-bit (execute/read, present, DPL=0, long mode)
        GDT[1] = GdtEntry::new(0x9A, 0x20);

        // Entry 2: Kernel Data 64-bit (read/write, present, DPL=0)
        GDT[2] = GdtEntry::new(0x92, 0x00);

        // Entry 3: User Code 64-bit (execute/read, present, DPL=3, long mode)
        GDT[3] = GdtEntry::new(0xFA, 0x20);

        // Entry 4: User Data 64-bit (read/write, present, DPL=3)
        GDT[4] = GdtEntry::new(0xF2, 0x00);

        // Entry 5-6: TSS (16 bytes, spans 2 GDT entries)
        let tss_addr = &TSS as *const Tss as u64;
        let tss_size = (core::mem::size_of::<Tss>() - 1) as u16;

        let tss_ptr = &mut GDT[5] as *mut GdtEntry as *mut TssEntry;
        (*tss_ptr) = TssEntry {
            length: tss_size,
            base_low: tss_addr as u16,
            base_mid: (tss_addr >> 16) as u8,
            flags1: 0x89,  // Present, 64-bit TSS (available)
            flags2: 0x00,
            base_high: (tss_addr >> 24) as u8,
            base_upper: (tss_addr >> 32) as u32,
            _reserved: 0,
        };

        // Load GDT
        GDTR = GdtDescriptor {
            limit: (core::mem::size_of::<[GdtEntry; GDT_ENTRIES]>() - 1) as u16,
            base: &GDT as *const _ as u64,
        };

        lgdt(&GDTR);

        // Reload segment registers
        core::arch::asm!(
            "mov ax, {data_sel:x}",
            "mov ds, ax",
            "mov es, ax",
            "mov fs, ax",
            "mov gs, ax",
            "mov ss, ax",
            data_sel = in(reg) KERNEL_DATA_SELECTOR as u64,
            options(nostack)
        );

        // Load TSS
        super::cpu::ltr(TSS_SELECTOR);
    }
}

/// Set the kernel stack pointer in TSS (used for Ring 3 → Ring 0 transitions)
pub fn set_kernel_stack(stack_top: u64) {
    unsafe {
        TSS.rsp0 = stack_top;
    }
}

File: FastOS_v2/kernel/gdt.c
/*
 * FastOS v2.0 — GDT (Global Descriptor Table) for 64-bit Long Mode
 * Reloads GDT in kernel space with proper TSS entry
 *
 * Segments:
 *   0x00: Null descriptor
 *   0x08: Kernel Code 64-bit (L=1, D=0)
 *   0x10: Kernel Data 64-bit
 *   0x18: User Code 64-bit (DPL=3)
 *   0x20: User Data 64-bit (DPL=3)
 *   0x28: TSS descriptor (16 bytes, spans 0x28-0x37)
 */

#include "include/kernel.h"

/* GDT entry: 8 bytes each */
typedef struct {
    uint16_t limit_low;
    uint16_t base_low;
    uint8_t  base_mid;
    uint8_t  access;
    uint8_t  granularity;
    uint8_t  base_high;
} __packed gdt_entry_t;

/* TSS for 64-bit Long Mode */
typedef struct {
    uint32_t reserved0;
    uint64_t rsp0;      /* Kernel stack for ring 0 */
    uint64_t rsp1;
    uint64_t rsp2;
    uint64_t reserved1;
    uint64_t ist1;      /* Interrupt Stack Table entries */
    uint64_t ist2;
    uint64_t ist3;
    uint64_t ist4;
    uint64_t ist5;
    uint64_t ist6;
    uint64_t ist7;
    uint64_t reserved2;
    uint16_t reserved3;
    uint16_t iomap_base;
} __packed tss_t;

/* TSS descriptor: 16 bytes (special in 64-bit mode) */
typedef struct {
    uint16_t limit_low;
    uint16_t base_low;
    uint8_t  base_mid;
    uint8_t  access;
    uint8_t  granularity;
    uint8_t  base_high;
    uint32_t base_upper;
    uint32_t reserved;
} __packed tss_desc_t;

/* GDT: 5 normal entries + 1 TSS (16 bytes) = 56 bytes */
static gdt_entry_t gdt[5] __aligned(16);
static tss_desc_t  gdt_tss __aligned(8);
static tss_t       tss __aligned(16);

static void gdt_set_entry(int idx, uint32_t base, uint32_t limit,
                           uint8_t access, uint8_t gran) {
    gdt[idx].limit_low   = limit & 0xFFFF;
    gdt[idx].base_low    = base & 0xFFFF;
    gdt[idx].base_mid    = (base >> 16) & 0xFF;
    gdt[idx].access      = access;
    gdt[idx].granularity  = ((limit >> 16) & 0x0F) | (gran & 0xF0);
    gdt[idx].base_high   = (base >> 24) & 0xFF;
}

static void gdt_set_tss(uint64_t base, uint32_t limit) {
    gdt_tss.limit_low   = limit & 0xFFFF;
    gdt_tss.base_low    = base & 0xFFFF;
    gdt_tss.base_mid    = (base >> 16) & 0xFF;
    gdt_tss.access      = 0x89;   /* Present, TSS available (64-bit) */
    gdt_tss.granularity  = ((limit >> 16) & 0x0F);
    gdt_tss.base_high   = (base >> 24) & 0xFF;
    gdt_tss.base_upper  = (uint32_t)(base >> 32);
    gdt_tss.reserved    = 0;
}

void gdt_init(void) {
    /* 0x00: Null */
    gdt_set_entry(0, 0, 0, 0, 0);

    /* 0x08: Kernel Code 64 — access=0x9A (Present,DPL0,Code,Exec,Read) gran=0x20 (Long mode) */
    gdt_set_entry(1, 0, 0, 0x9A, 0x20);

    /* 0x10: Kernel Data 64 — access=0x92 (Present,DPL0,Data,Write) */
    gdt_set_entry(2, 0, 0, 0x92, 0x00);

    /* 0x18: User Code 64 — access=0xFA (Present,DPL3,Code,Exec,Read) */
    gdt_set_entry(3, 0, 0, 0xFA, 0x20);

    /* 0x20: User Data 64 — access=0xF2 (Present,DPL3,Data,Write) */
    gdt_set_entry(4, 0, 0, 0xF2, 0x00);

    /* TSS: zero and set kernel stack */
    memset(&tss, 0, sizeof(tss));
    tss.rsp0 = 0x90000;    /* Kernel stack pointer */
    tss.iomap_base = sizeof(tss);

    /* 0x28: TSS descriptor */
    gdt_set_tss((uint64_t)&tss, sizeof(tss) - 1);

    /* Build GDTR — gdt is 5 entries (40 bytes) + tss_desc (16 bytes) = 56 bytes total
     * But they must be contiguous. We'll use a flat buffer approach. */
    static uint8_t gdt_flat[56] __aligned(16);
    memcpy(gdt_flat, gdt, sizeof(gdt));          /* 40 bytes */
    memcpy(gdt_flat + 40, &gdt_tss, 16);         /* 16 bytes */

    desc_ptr_t gdtr;
    gdtr.limit = sizeof(gdt_flat) - 1;
    gdtr.base  = (uint64_t)gdt_flat;
    lgdt(&gdtr);

    /* Reload data segments (CS is reloaded by retfq in asm, but we stay in ring 0) */
    hal_reload_segments();

    /* Load TSS (selector 0x28) */
    ltr(0x28);
}
